programs/token/src/state/mint.rs

```diff
@@ -0,0 +1,98 @@
+use crate::ID;
+
+use pinocchio::{
+    account_info::{AccountInfo, Ref},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+};
+pub struct Mint(*const u8);
+
+impl Mint {
+    pub const LEN: usize = 82;
+
+    /// Performs owner and length validation on `AccountInfo` and returns a `Ref<T>` for safe borrowing.
+    pub fn from_account_info(account_info: &AccountInfo) -> Result<Ref<Mint>, ProgramError> {
+        if account_info.data_len() != Self::LEN {
+            return Err(ProgramError::InvalidAccountData);
+        }
+        if account_info.owner() != &ID {
+            return Err(ProgramError::InvalidAccountData);
+        }
+        Ok(Ref::map(account_info.try_borrow_data()?, |data| unsafe {
+            &*(data.as_ptr() as *const Mint)
```

Comment:
Is this guaranteed to have the right alignment? otherwise it's UB

Reply:
Yes, the data comes from the `AccountInfo` (pointer from the entrypoint input), which is aligned.

Reply:
Uh I don't think that account pointers have a defined alignment, and the alignment isn't guaranteed to be mem::align_of::<Mint>() anyway?

Reply:
They should have, no?

https://solana.com/docs/programs/faq#input-parameter-serialization

Every account data starts from an 8-bytes aligned address. The `Mint` in this case is a struct that holds a pointer, so it should be aligned.

Reply:
Oh yes you're returning a ref to `Mint(ptr)` which is not the thing that is actually stored in the account, I missed that you had the pointer wrapper type